File: 1020. Number of Enclaves.java
class Solution {
    public int numEnclaves(int[][] grid) {
        int n=grid.length;
        int m=grid[0].length;

        int[][] vis=new int[n][m];
        
        for(int j=0;j<m;j++){
            if(grid[0][j]==1 && vis[0][j]==0) dfs(grid,vis,0,j);
        }
        for(int i=0;i<n;i++){
            if(grid[i][0]==1 && vis[i][0]==0) dfs(grid,vis,i,0);
        }
        for(int i=0;i<n;i++){
            if(grid[i][m-1]==1 && vis[i][m-1]==0) dfs(grid,vis,i,m-1);
        }
        for(int j=0;j<m;j++){
            if(grid[n-1][j]==1 && vis[n-1][j]==0) dfs(grid,vis,n-1,j);
        }
        
        int cnt=0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==1 && vis[i][j]==0){
                    cnt++;
                }
            }
        }
        
        return cnt;
    }
    
    public void dfs(int[][] a,int[][] vis,int i,int j){
        if(i<0 || i>=a.length || j<0 || j>=a[0].length || vis[i][j]==1 || a[i][j]==0){
            return ;
        }
        vis[i][j]=1;
        
        dfs(a,vis,i+1,j);
        dfs(a,vis,i-1,j);
        dfs(a,vis,i,j+1);
        dfs(a,vis,i,j-1);
        
        return;
    }
}

File: 1020-number-of-enclaves/1020-number-of-enclaves.cpp
class Solution {
public:
pair<int,bool> bfs(vector<vector<int>>& gr, vector<vector<bool>>& vis, int x, int y) {
	queue<pair<int, int>>qq;
	// (1,2)  (2,1)  (1,0)  (0,1)
	vector<int>xmoves = { 0,1,0,-1 };
	vector<int>ymoves = { 1,0,-1,0 };

	qq.push(make_pair(x, y));
	vis[x][y] = true;
	pair<int, int>temp;
	int r,c;
	bool is_at_boundary = false;
	int count = 0;
	while (!qq.empty()) {
		temp = qq.front();
		qq.pop();
		++count;
		
		for (int i = 0;i < 4;i++) {
			r = temp.first + xmoves[i];
			c = temp.second + ymoves[i];

			if (r < 0 || c < 0 || r >= gr.size() || c >= gr[0].size())
				is_at_boundary = true;
			else if (!vis[r][c] && gr[r][c]) {
				vis[r][c] = true;
				qq.push(make_pair(r, c));
			}

		}
	}

	return { count,is_at_boundary };
}
    int numEnclaves(vector<vector<int>>& grid) {
        int n = grid.size(), m = grid[0].size();
	int answer = 0;
	vector<vector<bool>>vis(n, vector<bool>(m,false));
	pair<int, bool>result;
	for (int i = 0;i < n;i++) {
		for (int j = 0;j < m;j++) {
			if (!vis[i][j] && grid[i][j]) {
				result = bfs(grid, vis, i, j);
				//cout <<"result is "<< result.first << endl;
				if (!result.second)answer += result.first;
			}
		}
	}

	return answer;
    }
};